discojs/src/models/gpt/layers.ts

```diff
@@ -210,81 +236,197 @@ class GELU extends tf.layers.Layer {
-    .apply(x1)
+  x2 = tf.layers.layerNormalization({
+      name: config.name + '.ln_2',
+      epsilon: 1e-5,
+      gammaInitializer: 'ones',
+      betaInitializer: 'zeros',
+  }).apply(x1)
+  if (config.debug) {
+    x2 = new LogLayer({ name: config.name + '.ln_2_log' }).apply(x2)
+  }
   
   // MLP 
-  x2 = MLP(Object.assign({}, config, { name: config.name })).apply(x2)
+  x2 = MLP(Object.assign({}, config, { name: config.name + '.mlp' })).apply(x2)
+  if (config.debug) {
+    x2 = new LogLayer({ name: config.name + '.mlp_log' }).apply(x2)
+  }
   // add attention output to mlp output
   x2 = tf.layers.add().apply([x1 as tf.SymbolicTensor, x2 as tf.SymbolicTensor])
+  if (config.debug) {
+    x2 = new LogLayer({ name: config.name + '.add_log' }).apply(x2)
+  }
 
   return tf.model({ name: config.name, inputs, outputs: x2 as tf.SymbolicTensor })
 }
 
 
+/**
+ * LanguageModelEmbedding is a layer that combines the token embeddings and the language modeling head
+ * I.e. LMEmbedding is used to translate token indices into token embeddings
+ * as well as to project embeddings back into token indices
+ * The GPT2 model uses the same embedding matrix for both the token embeddings and the language modeling head
+ * Because Tensorflow.js doesn't offer an easy weight sharing mechanism, we need to define a custom layer
+ * that can be used for both the token embeddings and the language modeling head.
+ * In the GPT2 model definition, this layers corresponds to wte and lm_head (which reuses wte)
+ */
+class LMEmbedding extends tf.layers.Layer {
+  static readonly className = 'LMEmbedding'
+  embeddings?: tf.LayerVariable
+
+  constructor(private readonly vocabSize: number,
+    private readonly nEmbd: number, private readonly seed: number) {
+    super({})
+  }
+  override build(): void {
+    this.embeddings = this.addWeight(
+      'wte', //use same name as GPT2
+      [this.vocabSize, this.nEmbd],
+      'float32',
+      tf.initializers.randomNormal({ mean:0, stddev:0.02, seed: this.seed })
+    )
+  }
+
+  override computeOutputShape(inputShape: tf.Shape | tf.Shape[]): tf.Shape | tf.Shape[] {
+    let shape: tf.Shape
+    if (Array.isArray(inputShape) && Array.isArray(inputShape[0])) shape = inputShape[0]
+    else shape = inputShape as tf.Shape
+    // input shape for the token embedding
+    if (shape.length === 2) {
+      // https://github.com/tensorflow/tfjs/blob/3daf152cb794f4da58fce5e21e09e8a4f89c8f80/tfjs-layers/src/layers/embeddings.ts#L155
+      // batch size and sequence length are undetermined 
+      // so the output shape is [null, null, nEmbd]
+      return [...shape, this.nEmbd]
+    }
+    // input shape for the language modeling head
+    // https://github.com/tensorflow/tfjs/blob/3daf152cb794f4da58fce5e21e09e8a4f89c8f80/tfjs-layers/src/layers/core.ts#L258
+    else if (shape.length === 3) {
+      shape[2] = this.vocabSize
+      return shape // [null, null, vocabSize]
```

Comment:
this reads more easily. (maybe you can also check before that `shape[0..1]` are indeed `null`.

```suggestion
      return [null, null, this.vocabSize]
```